// @vaadin/vaadin-element-mixin/vaadin-element-mixin.js
import { usageStatistics } from '@vaadin/vaadin-usage-statistics/vaadin-usage-statistics.js';
if (!window.Vaadin) {
  window['Vaadin'] = {};
}

/**
 * Array of Vaadin custom element classes that have been finalized.
 */
window['Vaadin'].registrations = window.Vaadin.registrations || [];

// Use the hack to prevent polymer-modulizer from converting to exports
window['Vaadin'].developmentModeCallback = window.Vaadin.developmentModeCallback || {};
window['Vaadin'].developmentModeCallback['vaadin-usage-statistics'] = function() {
  if (usageStatistics) {
    usageStatistics();
  }
};

/**
 * @polymerMixin
 */
export const ElementMixin = superClass => {
  try {
    return class VaadinElementMixin extends superClass {
      /** @protected */
      static _finalizeClass() {
        super._finalizeClass();

        // Registers a class prototype for telemetry purposes.
        if (this.is) {
          window.Vaadin.registrations.push(this);
        }
      }
      ready() {
        super.ready();
        if (document.doctype === null) {
          console.warn(
            'Vaadin components require the "standards mode" declaration. Please add <!DOCTYPE html> to the HTML document.'
          );
        }
      }
    };
  } finally {
    // This is run every time a new class is declared, not every time an instance is created
    if (window.Vaadin.developmentModeCallback) {
      window.Vaadin.developmentModeCallback['vaadin-usage-statistics']();
    }
  }
};
